torch/_functorch/_aot_autograd/subclass_utils.py

```diff
@@ -111,29 +122,73 @@ def create_subclass_meta(
 # a list of tensors that we would then need to concat together.
 # Instead, we specialize the logic for the inference vs. joint graph case.
 # NOTE: this function is hot, since we unwrap tensor subclass inputs at runtime
-def unwrap_tensor_subclasses(wrapped_args, *, is_joint_structure: bool):
-    def concat_inner_tensors_from_subclasses(xs):
-        xs_inner = []
-        for x in xs:
-            if is_traceable_wrapper_subclass(x):
-                xs_inner.extend(get_plain_tensors(typing.cast(Tensor, x)))
-            else:
-                xs_inner.append(x)
-        return xs_inner
 
-    if is_joint_structure:
-        assert isinstance(wrapped_args, tuple) and len(wrapped_args) == 2
-        assert isinstance(wrapped_args[0], (tuple, list)) and isinstance(
-            wrapped_args[1], (tuple, list)
-        )
-        unwrapped_args_fw = concat_inner_tensors_from_subclasses(wrapped_args[0])
-        unwrapped_args_tangents = concat_inner_tensors_from_subclasses(wrapped_args[1])
-        unwrapped_args = (unwrapped_args_fw, unwrapped_args_tangents)
-    else:
-        assert isinstance(wrapped_args, (list, tuple))
-        unwrapped_args_fw = concat_inner_tensors_from_subclasses(wrapped_args)
-        unwrapped_args = unwrapped_args_fw
-    return unwrapped_args
+
+def filter_symints(lst: Iterable[Union[int, SymInt]]) -> List[SymInt]:
+    return [s for s in lst if isinstance(s, SymInt) and not s.node.is_nested_int()]
+
+
+def compute_symint_placeholders(lst: Iterable[Union[int, SymInt]]) -> List[bool]:
+    return [isinstance(s, SymInt) and not s.node.is_nested_int() for s in lst]
+
+
+# The reason for "append_symints"
+#
+# * At compile time: we append extra symint args when unwrapping primals
+# (but not tangents, because they should always share symints with primals).
+# We also append extra symints when unwrapping the subclass outputs of the
+# traced function, so we can return them as extra outputs
+#
+# * At runtime: we similarly append subclass sizes when we unwrap subclass
+# primals (but not tangents) on entry to the forward.
+def unwrap_tensor_subclasses(
+    wrapped_args: List[Union[Tensor, int]],
+    *,
+    is_runtime: bool,
+    append_symints: bool,
+    subclass_metas: Optional[List[Union[int, SubclassCreationMeta]]] = None,
```

Comment:
Hmm, tbh this function is still a bit complicated. Two options that I think are reasonable (lmk which one you prefer):

(1) it is a bit silly that we have to pass in both an `is_runtime: bool` and a `subclass_metas: ...`, but the `subclass_metas` are only ever needed in the runtime case. So we could split off this function into a separate `unwrap_tensor_subclasses_compile_time` and `unwrap_tensor_subclasses_runtime`

(2) or alternatively, add a comment here describing how `subclass_metas` is used and why we need it (we only need it at runtime, because at runtime for every subclass graph input we need to decide which of its sizes/strides to pass in as extra dynamic args, and we get this info from the metadata)

Reply:
I'll go with (2) as `wrap_tensor_subclasses` also takes `is_runtime` as one of its arguments.

Reply:
As discussed offline, option (2) seems reasonable.